File: services/mutiny-svc/src/main/java/com/perfsim/mutinysvc/messaging/MessagingFaultService.java
package com.perfsim.mutinysvc.messaging;

import static java.nio.charset.StandardCharsets.UTF_8;

import com.rabbitmq.client.Channel;
import com.rabbitmq.client.Connection;
import com.rabbitmq.client.ConnectionFactory;
import io.opentelemetry.api.GlobalOpenTelemetry;
import io.opentelemetry.api.common.AttributeKey;
import io.opentelemetry.api.trace.Span;
import io.opentelemetry.api.trace.SpanKind;
import io.opentelemetry.api.trace.StatusCode;
import io.opentelemetry.api.trace.Tracer;
import io.opentelemetry.context.Scope;
import io.smallrye.mutiny.Uni;
import io.smallrye.mutiny.infrastructure.Infrastructure;
import jakarta.enterprise.context.ApplicationScoped;
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.TimeoutException;
import org.eclipse.microprofile.config.inject.ConfigProperty;

@ApplicationScoped
public class MessagingFaultService {

    private static final String EXCHANGE = "perfsim.mutiny-svc";
    private static final String QUEUE = "perfsim.mutiny-svc";
    private static final String ROUTING_KEY = "mutiny-svc";
    private static final long CONFIRM_TIMEOUT_MS = 5_000;
    private static final int NETWORK_MARGIN_MS = 5_000;

    private final String directHost;
    private final int directPort;
    private final String slowHost;
    private final int slowPort;
    private final String toxiproxyApi;
    private final Optional<String> username;
    private final Optional<String> password;
    private final HttpClient httpClient = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(5))
            .build();
    private final Tracer tracer = GlobalOpenTelemetry.getTracer(MessagingFaultService.class.getName());

    public MessagingFaultService(
            @ConfigProperty(name = "RABBITMQ_HOST", defaultValue = "rabbitmq.messaging.svc.cluster.local")
                    String directHost,
            @ConfigProperty(name = "RABBITMQ_PORT", defaultValue = "5672") int directPort,
            @ConfigProperty(name = "RABBITMQ_SLOW_HOST", defaultValue = "toxiproxy.messaging.svc.cluster.local")
                    String slowHost,
            @ConfigProperty(name = "RABBITMQ_SLOW_PORT", defaultValue = "25672") int slowPort,
            @ConfigProperty(name = "TOXIPROXY_API", defaultValue = "http://toxiproxy.messaging.svc.cluster.local:8474")
                    String toxiproxyApi,
            @ConfigProperty(name = "RABBITMQ_USERNAME") Optional<String> username,
            @ConfigProperty(name = "RABBITMQ_PASSWORD") Optional<String> password) {
        this.directHost = directHost;
        this.directPort = directPort;
        this.slowHost = slowHost;
        this.slowPort = slowPort;
        this.toxiproxyApi = toxiproxyApi;
        this.username = username;
        this.password = password;
    }

    public Uni<Map<String, Object>> publishSequentially(int messages) {
        return Uni.createFrom()
                .item(() -> publishSequentiallyBlocking(messages))
                .runSubscriptionOn(Infrastructure.getDefaultWorkerPool());
    }

    public Uni<Map<String, Object>> publishSlowly(long delayMs, int repeats) {
        return Uni.createFrom()
                .item(() -> publishSlowlyBlocking(delayMs, repeats))
                .runSubscriptionOn(Infrastructure.getDefaultWorkerPool());
    }

    private Map<String, Object> publishSequentiallyBlocking(int messages) {
        String configuredUsername = requireCredential("RABBITMQ_USERNAME", username);
        String configuredPassword = requireCredential("RABBITMQ_PASSWORD", password);
        try (Connection connection = newConnection(
                        directHost, directPort, 0, configuredUsername, configuredPassword);
                Channel channel = connection.createChannel()) {
            declareTopology(channel);
            channel.confirmSelect();
            for (int i = 0; i < messages; i++) {
                Span span = startProducerSpan();
                try (Scope ignored = span.makeCurrent()) {
                    channel.basicPublish(
                            EXCHANGE, ROUTING_KEY, null, ("mutiny-message-" + i).getBytes(UTF_8));
                } catch (Exception e) {
                    span.recordException(e);
                    span.setStatus(StatusCode.ERROR);
                    throw e;
                } finally {
                    span.end();
                }
            }
            channel.waitForConfirmsOrDie(CONFIRM_TIMEOUT_MS);
            return Map.of("published", messages, "confirmed", messages);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new IllegalStateException("Interrupted while awaiting RabbitMQ confirms", e);
        } catch (IOException | TimeoutException e) {
            throw new IllegalStateException("RabbitMQ publication failed", e);
        }
    }

    private Map<String, Object> publishSlowlyBlocking(long delayMs, int repeats) {
        String configuredUsername = requireCredential("RABBITMQ_USERNAME", username);
        String configuredPassword = requireCredential("RABBITMQ_PASSWORD", password);
        try {
            updateLatency(delayMs);
            int confirmed = 0;
            try (Connection connection = newConnection(
                            slowHost, slowPort, delayMs, configuredUsername, configuredPassword);
                    Channel channel = connection.createChannel()) {
                declareTopology(channel);
                channel.confirmSelect();
                for (int i = 0; i < repeats; i++) {
                    Span span = startProducerSpan();
                    try (Scope ignored = span.makeCurrent()) {
                        channel.basicPublish(
                                EXCHANGE,
                                ROUTING_KEY,
                                null,
                                ("slow-mutiny-message-" + i).getBytes(UTF_8));
                        channel.waitForConfirmsOrDie(delayMs + CONFIRM_TIMEOUT_MS);
                        confirmed++;
                    } catch (Exception e) {
                        span.recordException(e);
                        span.setStatus(StatusCode.ERROR);
                        throw e;
                    } finally {
                        span.end();
                    }
                }
            }
            return Map.of("published", repeats, "confirmed", confirmed, "delay_ms", delayMs);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new IllegalStateException("Interrupted while publishing through Toxiproxy", e);
        } catch (IOException | TimeoutException e) {
            throw new IllegalStateException("Slow RabbitMQ publication failed", e);
        }
    }

    private Span startProducerSpan() {
        return tracer.spanBuilder(EXCHANGE + " send")
                .setSpanKind(SpanKind.PRODUCER)
                .setAttribute(AttributeKey.stringKey("messaging.system"), "rabbitmq")
                .setAttribute(AttributeKey.stringKey("messaging.destination.name"), EXCHANGE)
                .setAttribute(AttributeKey.stringKey("messaging.operation.type"), "send")
                .startSpan();
    }

    private Connection newConnection(
            String host, int port, long downstreamDelayMs, String username, String password)
            throws IOException, TimeoutException {
        ConnectionFactory factory = new ConnectionFactory();
        factory.setHost(host);
        factory.setPort(port);
        factory.setUsername(username);
        factory.setPassword(password);
        int responseTimeoutMs = Math.toIntExact(downstreamDelayMs + NETWORK_MARGIN_MS);
        factory.setConnectionTimeout((int) CONFIRM_TIMEOUT_MS);
        factory.setHandshakeTimeout(downstreamDelayMs == 0
                ? (int) CONFIRM_TIMEOUT_MS
                : Math.toIntExact(2L * responseTimeoutMs + 1));
        factory.setChannelRpcTimeout(responseTimeoutMs);
        factory.setShutdownTimeout(responseTimeoutMs);
        return factory.newConnection();
    }

    private static String requireCredential(String name, Optional<String> value) {
        return value.filter(credential -> !credential.isBlank())
                .orElseThrow(() -> new IllegalStateException(name + " is required"));
    }

    private static void declareTopology(Channel channel) throws IOException {
        channel.exchangeDeclare(EXCHANGE, "direct", true);
        channel.queueDeclare(QUEUE, true, false, false, Map.of("x-message-ttl", 60_000));
        channel.queueBind(QUEUE, EXCHANGE, ROUTING_KEY);
    }

    private void updateLatency(long delayMs) throws IOException, InterruptedException {
        String updatePath = "/proxies/rabbitmq-slow/toxics/latency_downstream";
        String attributes = "{\"attributes\":{\"latency\":" + delayMs + ",\"jitter\":0}}";
        HttpResponse<Void> response = sendToxiproxy(updatePath, attributes);
        if (response.statusCode() == 404) {
            String createBody = "{\"name\":\"latency_downstream\","
                    + "\"type\":\"latency\",\"stream\":\"downstream\","
                    + "\"attributes\":{\"latency\":" + delayMs + ",\"jitter\":0}}";
            response = sendToxiproxy("/proxies/rabbitmq-slow/toxics", createBody);
            if (response.statusCode() == 409) {
                response = sendToxiproxy(updatePath, attributes);
            }
        }
        if (response.statusCode() < 200 || response.statusCode() >= 300) {
            throw new IOException("Toxiproxy returned HTTP " + response.statusCode());
        }
    }

    private HttpResponse<Void> sendToxiproxy(String path, String body)
            throws IOException, InterruptedException {
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(toxiproxyApi + path))
                .timeout(Duration.ofSeconds(5))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(body))
                .build();
        return httpClient.send(request, HttpResponse.BodyHandlers.discarding());
    }
}

File: services/rails-svc/app/services/messaging_fault_service.rb
require "bunny"
require "json"
require "net/http"
require "securerandom"
require "uri"

class MessagingFaultService
  DESTINATION = "perfsim.rails-svc".freeze
  ROUTING_KEY = "rails-svc".freeze
  DIRECT_TIMEOUT_SECONDS = 5

  def publish_sequentially(messages)
    confirmed = with_channel(slow: false, timeout: DIRECT_TIMEOUT_SECONDS) do |channel, exchange, returned|
      request_id = SecureRandom.uuid
      messages.times do |index|
        producer_span { publish(exchange, "rails-message-#{request_id}-#{index}") }
      end
      require_confirms(channel, returned)
      messages
    end

    { published: messages, confirmed: confirmed }
  end

  def publish_slowly(delay_ms, repeats)
    update_latency(delay_ms)
    operation_timeout = delay_ms / 1000.0 + DIRECT_TIMEOUT_SECONDS
    connection_timeout = 2 * operation_timeout + 1

    confirmed = with_channel(slow: true, timeout: connection_timeout, continuation_timeout: operation_timeout) do |channel, exchange, returned|
      request_id = SecureRandom.uuid
      repeats.times do |index|
        producer_span do
          publish(exchange, "slow-rails-message-#{request_id}-#{index}")
          require_confirms(channel, returned)
        end
      end
      repeats
    end

    { published: repeats, confirmed: confirmed, delay_ms: delay_ms }
  end

  private

  def with_channel(slow:, timeout:, continuation_timeout: timeout)
    username = required_env("RABBITMQ_USERNAME")
    password = required_env("RABBITMQ_PASSWORD")
    connection = Bunny.new(
      host: ENV.fetch(slow ? "RABBITMQ_SLOW_HOST" : "RABBITMQ_HOST",
                      slow ? "toxiproxy.messaging.svc.cluster.local" : "rabbitmq.messaging.svc.cluster.local"),
      port: Integer(ENV.fetch(slow ? "RABBITMQ_SLOW_PORT" : "RABBITMQ_PORT", slow ? "25672" : "5672"), 10),
      username: username,
      password: password,
      vhost: "/",
      automatically_recover: false,
      connection_timeout: timeout,
      read_timeout: timeout,
      write_timeout: timeout,
      continuation_timeout: (continuation_timeout * 1000).to_i,
    )
    channel = nil

    begin
      connection.start
      channel = connection.create_channel
      channel.confirm_select
      exchange = channel.direct(DESTINATION, durable: true, auto_delete: false)
      queue = channel.queue(
        DESTINATION,
        durable: true,
        auto_delete: false,
        arguments: { "x-message-ttl" => 60_000 },
      )
      queue.bind(exchange, routing_key: ROUTING_KEY)
      returned = []
      exchange.on_return { |info, _properties, _payload| returned << info }
      yield channel, exchange, returned
    ensure
      begin
        channel&.close if channel&.open?
      ensure
        connection.close
      end
    end
  end

  def publish(exchange, payload)
    exchange.publish(
      payload,
      routing_key: ROUTING_KEY,
      persistent: true,
      mandatory: true,
      content_type: "text/plain",
    )
  end

  def require_confirms(channel, returned)
    confirmed = channel.wait_for_confirms
    raise "RabbitMQ negatively acknowledged a publication" unless confirmed
    return if returned.empty?

    raise "RabbitMQ returned publication #{returned.first.reply_code}: #{returned.first.reply_text}"
  end

  def producer_span(&operation)
    OpenTelemetry.tracer_provider.tracer("rails-svc-messaging").in_span(
      "#{DESTINATION} send",
      kind: :producer,
      attributes: {
        "messaging.system" => "rabbitmq",
        "messaging.destination.name" => DESTINATION,
        "messaging.operation.type" => "send",
      },
      &operation
    )
  end

  def required_env(name)
    value = ENV[name]
    raise "#{name} is required" if value.nil? || value.empty?

    value
  end

  def update_latency(delay_ms)
    api = ENV.fetch("TOXIPROXY_API", "http://toxiproxy.messaging.svc.cluster.local:8474").delete_suffix("/")
    update = "#{api}/proxies/rabbitmq-slow/toxics/latency_downstream"
    status = post_json(update, attributes: { latency: delay_ms, jitter: 0 })
    if status == 404
      status = post_json("#{api}/proxies/rabbitmq-slow/toxics", {
        name: "latency_downstream",
        type: "latency",
        stream: "downstream",
        attributes: { latency: delay_ms, jitter: 0 },
      })
      status = post_json(update, attributes: { latency: delay_ms, jitter: 0 }) if status == 409
    end
    raise "Toxiproxy update failed with HTTP #{status}" unless status.between?(200, 299)
  end

  def post_json(url, body)
    uri = URI(url)
    request = Net::HTTP::Post.new(uri)
    request["Content-Type"] = "application/json"
    request.body = JSON.generate(body)
    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = uri.scheme == "https"
    http.open_timeout = DIRECT_TIMEOUT_SECONDS
    http.read_timeout = DIRECT_TIMEOUT_SECONDS
    http.write_timeout = DIRECT_TIMEOUT_SECONDS
    http.start { |client| client.request(request) }.code.to_i
  end
end
